File: frontend/src/api/queries/firewall.ts
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query';
import { HttpUtil } from '../http-init';

export interface FirewallRule {
  id: number;
  ordering: number;
  action: 'accept' | 'drop' | 'reject';
  proto: 'tcp' | 'udp' | 'any';
  port_spec: string;
  source_cidr: string;
  comment: string;
  enabled: boolean;
}

export interface FirewallBaselinePort {
  proto: string;
  port: number;
  label: string;
}

export interface FirewallPolicy {
  enabled: boolean;
  default_incoming: 'drop' | 'accept';
  rollback_window_secs: number;
  last_applied_at?: string;
  last_confirmed_at?: string;
}

export interface FirewallStatus {
  pending: boolean;
  remaining_secs: number;
  confirmed_state_saved: boolean;
}

export interface FirewallGetResp {
  policy: FirewallPolicy;
  rules: FirewallRule[];
  baseline: FirewallBaselinePort[];
  warnings: string[];
  status?: FirewallStatus;
}

export function useFirewallQuery(serverId: string) {
  return useQuery({
    queryKey: ['firewall', serverId],
    queryFn: () => HttpUtil.get<FirewallGetResp>(`/api/servers/${encodeURIComponent(serverId)}/firewall`),
    enabled: !!serverId,
  });
}

export function useFirewallStatusQuery(serverId: string, enabled: boolean) {
  return useQuery({
    queryKey: ['firewall', serverId, 'status'],
    queryFn: () => HttpUtil.get<FirewallStatus>(`/api/servers/${encodeURIComponent(serverId)}/firewall/status`),
    enabled,
    refetchInterval: enabled ? 2000 : false,
  });
}

export interface FirewallRuleInput {
  action: 'accept' | 'drop' | 'reject';
  proto: 'tcp' | 'udp' | 'any';
  port_spec: string;
  source_cidr: string;
  comment: string;
  enabled: boolean;
}

export interface FirewallPolicyInput {
  enabled: boolean;
  default_incoming: 'drop' | 'accept';
  rollback_window_secs: number;
}

export interface FirewallDryRunResp {
  valid: boolean;
  error?: string;
  added: string[];
  removed: string[];
}

export interface FirewallApplyResp {
  rollback_window_secs: number;
  panel_reachable: boolean;
}

export function useFirewallMutations(serverId: string) {
  const qc = useQueryClient();
  const invalidate = () => void qc.invalidateQueries({ queryKey: ['firewall', serverId] });

  const savePolicy = useMutation({
    mutationFn: (input: FirewallPolicyInput) => HttpUtil.put<null>(`/api/servers/${encodeURIComponent(serverId)}/firewall/policy`, input),
    onSuccess: invalidate,
  });
  const saveRules = useMutation({
    mutationFn: (rules: FirewallRuleInput[]) => HttpUtil.put<null>(`/api/servers/${encodeURIComponent(serverId)}/firewall/rules`, rules),
    onSuccess: invalidate,
  });
  const dryRun = useMutation({
    mutationFn: () => HttpUtil.post<FirewallDryRunResp>(`/api/servers/${encodeURIComponent(serverId)}/firewall/dry-run`),
  });
  const apply = useMutation({
    mutationFn: () => HttpUtil.post<FirewallApplyResp>(`/api/servers/${encodeURIComponent(serverId)}/firewall/apply`),
    onSuccess: invalidate,
  });
  const confirm = useMutation({
    mutationFn: () => HttpUtil.post<null>(`/api/servers/${encodeURIComponent(serverId)}/firewall/confirm`),
    onSuccess: invalidate,
  });
  const rollback = useMutation({
    mutationFn: () => HttpUtil.post<null>(`/api/servers/${encodeURIComponent(serverId)}/firewall/rollback`),
    onSuccess: invalidate,
  });

  return { savePolicy, saveRules, dryRun, apply, confirm, rollback };
}

File: internal/store/firewall.go
package store

import (
	"context"
	"errors"
	"time"

	"github.com/jackc/pgx/v5"
)

// FirewallPolicy is one server's INPUT-chain firewall settings. Baseline
// "never lock these out" ports are NOT part of this struct -- they're
// computed fresh at every apply from live DB/host state (see
// internal/firewall's renderer), never persisted here.
type FirewallPolicy struct {
	ServerID           string
	Enabled            bool
	DefaultIncoming    string // "drop" | "accept"
	RollbackWindowSecs int
	LastAppliedRuleset string
	LastAppliedAt      *time.Time
	LastConfirmedAt    *time.Time
	UpdatedAt          time.Time
}

// defaultFirewallPolicy is what a server with no row yet effectively has:
// disabled, drop-by-default once enabled, a 5-minute rollback window.
func defaultFirewallPolicy(serverID string) FirewallPolicy {
	return FirewallPolicy{ServerID: serverID, DefaultIncoming: "drop", RollbackWindowSecs: 300}
}

// GetFirewallPolicy returns serverID's policy, or sensible defaults (not an
// error) if it hasn't configured one yet -- "no row" is a normal state for
// a server that hasn't opted into this feature, not a failure.
func (s *Store) GetFirewallPolicy(ctx context.Context, serverID string) (FirewallPolicy, error) {
	var p FirewallPolicy
	err := s.pool.QueryRow(ctx, `
		SELECT server_id, enabled, default_incoming, rollback_window_secs,
		       last_applied_ruleset, last_applied_at, last_confirmed_at, updated_at
		FROM protean.firewall_policy WHERE server_id = $1
	`, serverID).Scan(&p.ServerID, &p.Enabled, &p.DefaultIncoming, &p.RollbackWindowSecs,
		&p.LastAppliedRuleset, &p.LastAppliedAt, &p.LastConfirmedAt, &p.UpdatedAt)
	if errors.Is(err, pgx.ErrNoRows) {
		return defaultFirewallPolicy(serverID), nil
	}
	return p, err
}

// UpsertFirewallPolicy saves the admin-editable policy fields (enabled,
// default_incoming, rollback_window_secs) -- last_applied_*/last_confirmed_*
// are only ever set by SetLastApplied/SetLastConfirmed, reflecting an
// actual apply/confirm, never a plain settings edit.
func (s *Store) UpsertFirewallPolicy(ctx context.Context, p FirewallPolicy) error {
	_, err := s.pool.Exec(ctx, `
		INSERT INTO protean.firewall_policy (server_id, enabled, default_incoming, rollback_window_secs)
		VALUES ($1, $2, $3, $4)
		ON CONFLICT (server_id) DO UPDATE SET
			enabled = $2, default_incoming = $3, rollback_window_secs = $4, updated_at = now()
	`, p.ServerID, p.Enabled, p.DefaultIncoming, p.RollbackWindowSecs)
	return err
}

// SetLastApplied records a ruleset actually pushed to the host (armed
// rollback pending, not yet confirmed).
func (s *Store) SetLastApplied(ctx context.Context, serverID, ruleset string, at time.Time) error {
	_, err := s.pool.Exec(ctx, `
		INSERT INTO protean.firewall_policy (server_id, last_applied_ruleset, last_applied_at)
		VALUES ($1, $2, $3)
		ON CONFLICT (server_id) DO UPDATE SET
			last_applied_ruleset = $2, last_applied_at = $3, updated_at = now()
	`, serverID, ruleset, at)
	return err
}

// SetLastConfirmed records that firewall-confirm succeeded -- the ruleset
// in last_applied_ruleset is now the persisted, boot-surviving one.
func (s *Store) SetLastConfirmed(ctx context.Context, serverID string, at time.Time) error {
	_, err := s.pool.Exec(ctx, `
		UPDATE protean.firewall_policy SET last_confirmed_at = $2, updated_at = now() WHERE server_id = $1
	`, serverID, at)
	return err
}

// FirewallRule is one of the admin's own custom INPUT rules, evaluated
// (in Ordering order) after the baseline/loopback/established rules the
// renderer always prepends.
type FirewallRule struct {
	ID         int64
	ServerID   string
	Ordering   int
	Action     string // accept|drop|reject
	Proto      string // tcp|udp|any
	PortSpec   string // "443", "8000:8100", "80,443" -- empty = any port
	SourceCIDR string // empty = anywhere
	Comment    string
	Enabled    bool
	CreatedAt  time.Time
}

func (s *Store) ListFirewallRules(ctx context.Context, serverID string) ([]FirewallRule, error) {
	rows, err := s.pool.Query(ctx, `
		SELECT id, server_id, ordering, action, proto, port_spec, source_cidr, comment, enabled, created_at
		FROM protean.firewall_rules WHERE server_id = $1 ORDER BY ordering
	`, serverID)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var out []FirewallRule
	for rows.Next() {
		var r FirewallRule
		if err := rows.Scan(&r.ID, &r.ServerID, &r.Ordering, &r.Action, &r.Proto,
			&r.PortSpec, &r.SourceCIDR, &r.Comment, &r.Enabled, &r.CreatedAt); err != nil {
			return nil, err
		}
		out = append(out, r)
	}
	return out, rows.Err()
}

// ReplaceFirewallRules atomically swaps serverID's entire custom rule set
// -- the draft-editing UI always sends the full set, not incremental
// add/remove, so a transactional delete+bulk-insert keeps this simple and
// leaves no orphaned rows on a partial failure.
func (s *Store) ReplaceFirewallRules(ctx context.Context, serverID string, rules []FirewallRule) error {
	tx, err := s.pool.Begin(ctx)
	if err != nil {
		return err
	}
	defer tx.Rollback(ctx)
	if _, err := tx.Exec(ctx, `DELETE FROM protean.firewall_rules WHERE server_id = $1`, serverID); err != nil {
		return err
	}
	for _, r := range rules {
		if _, err := tx.Exec(ctx, `
			INSERT INTO protean.firewall_rules (server_id, ordering, action, proto, port_spec, source_cidr, comment, enabled)
			VALUES ($1, $2, $3, $4, $5, $6, $7, $8)
		`, serverID, r.Ordering, r.Action, r.Proto, r.PortSpec, r.SourceCIDR, r.Comment, r.Enabled); err != nil {
			return err
		}
	}
	return tx.Commit(ctx)
}
